cloudinit/net/dhcp.py

```diff
@@ -532,22 +530,132 @@ def parse_dhcp_server_from_lease_file(lease_file) -> Optional[str]:
         return latest_address
 
 
-class Dhcpcd:
+class Dhcpcd(DhcpClient):
     client_name = "dhcpcd"
 
-    def __init__(self):
-        raise NoDHCPLeaseMissingDhclientError("Dhcpcd not yet implemented")
+    def dhcp_discovery(
+        self,
+        interface,
+        dhcp_log_func=None,
+        distro=None,
+    ):
+        """Run dhclient on the interface without scripts/filesystem artifacts.
+
+        @param dhclient_cmd_path: Full path to the dhclient used.
+        @param interface: Name of the network interface on which to dhclient.
+        @param dhcp_log_func: A callable accepting the dhclient output and
+            error streams.
+
+        @return: A list of dicts of representing the dhcp leases parsed from
+            the dhclient.lease file or empty list.
+        """
+        LOG.debug("Performing a dhcp discovery on %s", interface)
+
+        # ISC dhclient needs the interface up to send initial discovery packets
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        distro.net_ops.link_up(interface)
+
+        # TODO: disabling hooks means we need to get all of the files in
+        # /lib/dhcpcd/dhcpcd-hooks/ and pass each of those with the --nohook
+        # argument to dhcpcd
+        try:
+            subp.subp(
+                [
+                    self.client_name,
+                    "--oneshot",  # get lease then exit
+                    "--nobackground",  # don't fork
+                    "--ipv4only",  # only attempt configuring ipv4
+                    "--waitip=4",  # wait for ipv4 to be configured
+                    "--persistent",  # don't deconfigure when dhcpcd exits
+                    "--noarp",  # don't be slow
+                    interface,
+                ]
+            )
+            return self.parse_dhcp_lease_file(interface)
+        except subp.ProcessExecutionError as error:
+            LOG.debug(
+                "dhclient exited with code: %s stderr: %r stdout: %r",
```

Comment:
```suggestion
                f"{self.client_name} exited with code: %s stderr: %r stdout: %r",
```

Reply:
why do that? this is in dhclient-specific code

Reply:
good pt unnecessary. Was just using my hammer throughout logging comments